optuna/storages/_journal/storage.py

```diff
@@ -0,0 +1,608 @@
+            else:
+                log["datetime_start"] = None
+            if template_trial.datetime_complete:
+                log["datetime_complete"] = template_trial.datetime_complete.isoformat()
+
+            log["distributions"] = {
+                k: distribution_to_json(dist) for k, dist in template_trial.distributions.items()
+            }
+            log["params"] = {
+                k: template_trial.distributions[k].to_internal_repr(param)
+                for k, param in template_trial.params.items()
+            }
+            log["user_attrs"] = template_trial.user_attrs
+            log["system_attrs"] = template_trial.system_attrs
+            log["intermediate_values"] = template_trial.intermediate_values
+
+        with self._thread_lock:
+            self._write_log(JournalOperation.CREATE_TRIAL, log)
+            self._sync_with_backend()
+            return self._replay_result._trial_ids_owned_by_this_process[-1]
+
+    def set_trial_param(
+        self,
+        trial_id: int,
+        param_name: str,
+        param_value_internal: float,
+        distribution: BaseDistribution,
+    ) -> None:
+        log: Dict[str, Any] = {
+            "trial_id": trial_id,
+            "param_name": param_name,
+            "param_value_internal": param_value_internal,
+            "distribution": distribution_to_json(distribution),
+        }
+
+        with self._thread_lock:
+            self._write_log(JournalOperation.SET_TRIAL_PARAM, log)
+            self._sync_with_backend()
+
+    def get_trial_id_from_study_id_trial_number(self, study_id: int, trial_number: int) -> int:
+        with self._thread_lock:
+            self._sync_with_backend()
+            if len(self._replay_result._study_id_to_trial_ids[study_id]) <= trial_number:
+                raise KeyError(
+                    "No trial with trial number {} exists in study with study_id {}.".format(
+                        trial_number, study_id
+                    )
+                )
+            return self._replay_result._study_id_to_trial_ids[study_id][trial_number]
+
+    def set_trial_state_values(
+        self, trial_id: int, state: TrialState, values: Optional[Sequence[float]] = None
+    ) -> bool:
+        log: Dict[str, Any] = {
+            "trial_id": trial_id,
+            "state": state,
+            "values": values,
+        }
+
+        if state == TrialState.RUNNING:
+            log["datetime_start"] = datetime.datetime.now().isoformat()
+        elif state.is_finished():
+            log["datetime_complete"] = datetime.datetime.now().isoformat()
+
+        with self._thread_lock:
+            self._write_log(JournalOperation.SET_TRIAL_STATE_VALUES, log)
+            self._sync_with_backend()
+
+            if (
+                state == TrialState.RUNNING
+                and trial_id not in self._replay_result._trial_ids_owned_by_this_process
+            ):
+                return False
```

Comment:
[question] What happens when the trial state becomes `WAITING` to `RUNNING` in the multi-processing situation?

I will pick up a concrete example like:
1. `The process A` enqueues `the trial 1` with `WAITING` state. In that case, `the process A` has `the trial 1` in `_trial_ids_owned_by_this_process`.
2. And, as it is in the multi-processing situation, `the process B` asks `the trial 1` coincidentally. In that case, `the process B` tries to `set_trial_state_values(trial._trial_id, state=TrialState.RUNNING)` into `the trial 1`.
3. However, `the process B` does not have `the trial 1` in the `_trial_ids_owned_by_this_process`. So, `set_trial_state_values` returns `False`. It is different behavior comparing to the current master, AFAIK.

It is a complicated behavior, and I might be wrong, but could you check it?

Reply:
And, if it is a bug, we need to add a unit test for it. Also, we need to carefully check that we have enough unit tests around `WAITING` state behaviors.

Reply:
Thank you for reporting the case. The implementation was wrong. Fixed in [b67e0b5](https://github.com/optuna/optuna/pull/3854/commits/b67e0b53d4aa33ab0e0bc64a96942732deb1c0c6).